### DISCUSSION TITLE
A fallback replacement_scanner for httpfs

### DISCUSSION THREAD

**Discussion by douenergy** [duckdb/duckdb#8174] [General]
## Proposal: Adding `httpfs_fallback_format` Option

```sql
FROM read_json_auto('httpfs://some-api-endpoint')
```
to 

```sql
FROM 'httpfs://some-api-endpoint' 
```

### Description
This proposal suggests adding the `httpfs_fallback_format` option. The purpose of this option is to allow users to specify a fallback format when querying web url that do not have the `.json` string in the URL path.

### Motivation
Currently, when using the `httpfs` extension to query web APIs, it requires manual handling for APIs that do not have the `.json` string in the URL path. This involves using functions like `read_json_auto` or `read_json` with the API URL. By introducing the `httpfs_fallback_format` option, we can simplify the process.

The JSON format is widely used in web APIs, such as the GitHub API and Stack Overflow API. Many developers are already familiar with analyzing JSON data from these APIs. By defaulting the `httpfs_fallback_format` to JSON (or optionally `None`), we can automatically treat web APIs without the `.json` string as JSON data, making the process of analyzing web APIs more convenient.

I believe the core idea behind the replacement scan is to enhance the convenience of common paths. Could httpfs_fallback_format potentially contribute to a more seamless experience when querying web APIs?

If we think this is good to have. We can discuss a few points:

Naming: Should we use `httpfs_fallback_format` or `httpfs_default_format` as the name for the option? This depends on the desired clarity and naming convention in the code. If the purpose is to specify the default format for web API queries, `httpfs_default_format` might be a more appropriate name. On the other hand, if the intention is to indicate that this format will be used as a fallback when the .json string is not present in the URL, `httpfs_fallback_format` can provide a more descriptive name.

Default value: Should we set the default value as "None" instead of "JSON"? Setting the default value as "None" gives users the flexibility to explicitly specify the desired format when querying web APIs, without making any assumptions. This is particularly useful in scenarios where the format may vary or when querying non-JSON APIs. It's a good suggestion to consider making the default value configurable or set it as "None".

### Implementation:
To implement this feature, we can consider adding a replacement_scanner to config.replacement_scans. This ensures that all existing replacement scans will continue to function as before. However, when querying from web APIs, we can make use of the httpfs_fallback_format replacement scanner as a fallback option. This approach maintains compatibility with the current functionality while applying the fallback format when necessary.

@Mytherin @samansmink @lnkuiper @MarkRoddy

**Comment by Tishj** [comment]
Would love to see an example of this option, it's not entirely clear what you propose this should look like.

Would it be something like:
```sql
SET httpfs_fallback_format='json';
```
?

**Comment by douenergy** [comment]
Thanks @Tishj , let me explain more clearly.

There are only 4 valid setting values for httpfs_fallback_format:

* csv
* parquet
* json
* none

No matter what value httpfs_fallback_format is, it must not affect the existing replacement scan.

For example:

```sql
/* No matter how httpfs_fallback_format changes, everything below should work. */

FROM 'local/some-csv-data.csv' // work
FROM 'https://some-parquet-data.parquet' // work

--- python ---

df = dffunction()
duckdb.sql('FROM df') // work
.
.
.
```

If we set httpfs_fallback_format to None, that should work as DuckDB does now.
Otherwise, we will add a fallback replacement scan, which can be one of csv, parquet, or json(control by the value of httpfs_fallback_format).

```sql
/*'https://some-api-endpoint  -- > json */
/*'https://some-parquet-endpoint.parquet  -- > parquet */
/*'https://some-parquet-endpoint  -- > parquet */

SET httpfs_fallback_format='none';

FROM read_json_auto('https://some-api-endpoint') // work
FROM 'https://some-api-endpoint' // error ❌

--- json ---
SET httpfs_fallback_format='json';

FROM read_json_auto('https://some-api-endpoint') // work
FROM 'https://some-api-endpoint' // work

FROM 'https://some-parquet-endpoint.parquet' // work (parquet replacement scan)
FROM 'https://some-parquet-endpoint' // error ❌ (use json reader to read the parquet)

--- parquet ---
SET httpfs_fallback_format='parquet';

FROM read_parquet('https://some-parquet-endpoint') // work
FROM 'https://some-parquet-endpoint' // work

FROM 'https://some-api-endpoint' // error ❌ (use parquet reader to read json)

```

The idea is to add one of read_csv_auto, read_json_auto, or read_parquet to every URL that doesn't have a file extension.
This is only for httpfs.
